File: src/lexer.py
from enum import Enum, auto 
from typing import Optional
from util import eprint

class TokenKind(Enum):
    PLUS = auto(),
    DASH = auto(),
    ASTERISK = auto()
    SLASH_FORWARD = auto(),
    NEWLINE = auto(),
    EOF = auto()
    LITERAL_INT = auto(),
    LITERAL_FLOAT = auto(),
    LITERAL_STRING = auto(),
    LITERAL_BOOL = auto(),
    DOUBLE_EQUAL = auto()
    NOT_EQUAL = auto()
    GREATER_THAN = auto()
    LESS_THAN = auto()
    LESS_THAN_EQUAL = auto()
    GREATER_THAN_EQUAL = auto()
    DOT = auto(),
    SEMICOLON = auto()
    COMMA = auto()
    EXCLAMATION_MARK = auto()
    LPAREN = auto(),
    RPAREN = auto(),
    SINGLE_EQUAL = auto(),
    PROC = auto(),
    DO = auto()
    HAT = auto()
    SLIM_ARROW = auto()
    TYPE_ = auto()
    COLON = auto()
    WHILE = auto()
    IF = auto()
    END = auto()
    ELSE = auto()
    MEMORY = auto()
    RETURN = auto()
    EXTERN = auto()
    PUB = auto()
    IDENTIFIER = auto()


class Token:
    def __init__(self, kind:TokenKind, value:str, row:int, col:int, filename:str):
        self.kind = kind
        self.value = value
        self.row = row
        self.col = col
        self.file = filename
    
    def __repr__(self):
         name_or_kind = self.kind.name
         if self.value != "" and not self.value.isspace():
            name_or_kind = self.value
         rep = f"{name_or_kind} in {self.file} at line: {self.row+1} and column: {self.col+1}"
         return rep


class Lexer:
    def __init__(self, code: str, filename:str):
        self.code = code
        self.pos = 0
        self.row = 0
        self.col = 0
        self.filename = filename
    
    #returns the current character pointed to by the lexer.
    def __char(self) -> str:
        self.__consume()
        return self.code[self.pos-1]
    
    #advances the lexer 
    def __consume(self):
        self.pos += 1
        self.col += 1
    
    #returns the current character without advancing the lexer
    #returns none if we reached the end of the file
    def __peek_char(self) -> Optional[str]:
        if self.pos > len(self.code)-1: return None
        return self.code[self.pos]
    
    
    #advances the lexer until there is no whitespace returns true if reached end of file
    def __skip_whitespace(self) -> bool:
        while True:
            if self.__peek_char():
                if self.__peek_char().isspace():
                   self.__consume()
                else:
                    return False
            else:
                return True
    
    #return every character collected until the filter function returns false
    def __get_characters_while(self, filter_function) -> str:
        characters = ""
        while True:
            if self.__peek_char() == None:
                return characters
            if not filter_function(self.__peek_char()):
                return characters
            characters += self.__char()
    
    #returns the next token returns EOF token if we reached the end.
    def next(self) -> Token:
        #since newline counts as whitespace this mandatory to check before skipping whitespace
        if self.__peek_char():
            if self.__peek_char() == "\n":
                self.__consume()
                self.row += 1
                self.col = 0
                return Token(TokenKind.NEWLINE, "\n", self.row, self.col, self.filename)
        else:
            return Token(TokenKind.EOF, "", self.row, self.col, self.filename)
        
        #skip to first character of next token
        eof = self.__skip_whitespace()
        if eof: return Token(TokenKind.EOF, "", self.row, self.col, self.filename)

        #create our token object
        token = Token(None, "", self.row, self.col, self.filename)
        
        #get the first character of our token
        first_character = self.__char()

        token.value += first_character

        match first_character:
            case "+": token.kind = TokenKind.PLUS
            case "-": token.kind = TokenKind.DASH
            case "*": token.kind = TokenKind.ASTERISK
            case "/": token.kind = TokenKind.SLASH_FORWARD
            case "#":
                _ = self.__get_characters_while(lambda x: x != "\n")
                if self.__peek_char() == None:
                    token.kind = TokenKind.EOF
                else:
                    return self.next()
            case "\"":
                token.kind = TokenKind.LITERAL_STRING
                token.value = self.__get_characters_while(lambda x: x != "\"")
                if self.__peek_char() == None:
                    eprint(f"Expected \" to end string literal started at row: {token.row+1} col: {token.col+1}")
                self.__consume()
            
            case ".": token.kind = TokenKind.DOT
            case "(": token.kind = TokenKind.LPAREN
            case ")": token.kind = TokenKind.RPAREN
            case ";": token.kind = TokenKind.SEMICOLON
            case ",": token.kind = TokenKind.COMMA
            case "^": token.kind = TokenKind.HAT
            case ":": token.kind = TokenKind.COLON
            case ";": token.kind = TokenKind.SEMICOLON
            case "=":
                if self.__peek_char() == None: eprint( f"cannot end file on {first_character}")
                if self.__peek_char() == "=":
                    self.__consume()
                    token.kind = TokenKind.DOUBLE_EQUAL
                else: token.kind = TokenKind.SINGLE_EQUAL
            
            case ">":
                if self.__peek_char() == None: eprint( f"cannot end file on {first_character}")
                if self.__peek_char() == "=":
                    self.__consume()
                    token.kind = TokenKind.GREATER_THAN_EQUAL
                else: token.kind = TokenKind.GREATER_THAN

            case "<":
                if self.__peek_char() == None: eprint( f"cannot end file on {first_character}")
                if self.__peek_char() == "=":
                    self.__consume()
                    token.kind = TokenKind.LESS_THAN_EQUAL
                elif self.__peek_char() == "-":
                    self.__consume()
                    token.kind = TokenKind.SLIM_ARROW
                else: token.kind = TokenKind.LESS_THAN
            
            case "!":
                if self.__peek_char() == None: eprint( f"cannot end file on {first_character}")
                if self.__peek_char() == "=":
                    self.__consume()
                    token.kind = TokenKind.NOT_EQUAL
                else: token.kind = TokenKind.EXCLAMATION_MARK
            #this would be a keyword, identifier or an integer/float
            case _:
                value = first_character
                value += self.__get_characters_while(lambda x: x.isalnum() or x == "_")
                if value.isnumeric():
                    if self.__peek_char():
                        if self.__peek_char() != ".":
                            token.kind = TokenKind.LITERAL_INT
                        else:
                            token.kind = TokenKind.LITERAL_FLOAT
                            value += self.__char()
                            rest = self.__get_characters_while(lambda x: x.isnumeric())
                            value += rest
                else:
                    match value:
                        case "proc": token.kind = TokenKind.PROC
                        case "do": token.kind = TokenKind.DO
                        case "if": token.kind= TokenKind.IF
                        case "else": token.kind = TokenKind.ELSE
                        case "while": token.kind = TokenKind.WHILE
                        case "end": token.kind = TokenKind.END
                        case "true" | "false": token.kind = TokenKind.LITERAL_BOOL
                        case "return": token.kind = TokenKind.RETURN
                        case "memory": token.kind = TokenKind.MEMORY
                        case "extern": token.kind = TokenKind.EXTERN
                        case "pub": token.kind = TokenKind.PUB
                        case _: token.kind = TokenKind.IDENTIFIER
                token.value = value

        
        return token

    def peek_next_token(self) -> Token:
        cur_pos = self.pos
        cur_row = self.row
        cur_col = self.col
        tk = self.next()
        self.pos = cur_pos
        self.row = cur_row
        self.col = cur_col
        return tk

        
        
        

File: src/lexer.ts
export enum TokenType{
    LiteralInt,
    LiteralFloat,
    LiteralBool,
    LiteralString,
    Identifier,
    Plus,
    Dash,
    SlashForward,
    Asterisk,
    LessThan,
    LessThanEqual,
    GreaterThan,
    GreaterThanEqual,
    SingeEqual,
    DoubleEqual,
    NotEqual,
    ExclamationMark,
    Lparen,
    Rparen,
    Semicolon,
    Colon,
    Comma,
    Dot,
    Hat,
    If,
    Do,
    While,
    Proc,
    End,
    Return,
    Memory,
    Alloc,
    Struct,
    Const,
    Pub,
    Newline,
    EOF,
    Environment,
}

export interface Token{
    kind: TokenType,
    value: string,
    row: number,
    col: number,
    filename: string,

}

export const token_repr = function(t: Token): string {
    const type = TokenType[t.kind];
    let repr = `${t.value} at line: ${t.row+1} col: ${t.col+1}\n`
    repr += `   The token was parsed as type ${type}`
    return repr;
}

type FilterFunction = (x: string) => boolean;

const keywords = new Map<string, TokenType>([
    ["if", TokenType.If],
    ["do", TokenType.Do],
    ["while", TokenType.While],
    ["end", TokenType.End],
    ["proc", TokenType.Proc],
    ["return", TokenType.Return],
    ["pub", TokenType.Pub],
    ["memory", TokenType.Memory],
    ["alloc", TokenType.Alloc],
    ["struct", TokenType.Struct],
    ["const", TokenType.Const],
    ["environment", TokenType.Environment],

])


export class Lexer{
    filestring: string
    filename: string
    pos: number
    row: number
    col: number
    constructor(filestring: string, filename: string){
        this.filestring = filestring;
        this.filename = filename;
        this.pos = 0;
        this.row = 0;
        this.col = 0;
    }
    
    //returns a token based on current state of the lexer
    private newToken(initial_value: string, type?:TokenType): Token{
        return {
            kind: type ?? TokenType.Identifier,
            value:initial_value, 
            row: this.row, col: this.col, 
            filename: this.filename
        };
    }

    //returns current character pointed to by the lexer and moves on
    private char(): string{
        const c = this.filestring[this.pos];
        this.advance();
        return c;

    }
    
    //returns current character but does not advance the lexer
    private peek(): string | null {
        if(this.pos > this.filestring.length-1) return null;
        return this.filestring[this.pos];
    }

    //moves the lexer to the next position
    private advance() {
        this.pos += 1
        this.col += 1
    }

    //skips until the first non-whitespace character return result indicates if at eof
    private skipWhitespace(): boolean{
        while(true){
            if(!this.peek()) return true;
            if(!/\s/.test(this.peek() as string)) return false;
            this.advance()
        }

    }

    //while the provided function is true collect characters
    private get_characters_while(filter_function: FilterFunction): Array<string>{
        const result = []
        while(true){
            if(this.peek() == null || !filter_function(this.peek() as string)) break;
            result.push(this.char())
        }
        return result;
    }



    next(): Token{

        if(this.peek()){
            if(this.peek() == "\n"){
                this.advance();
                this.row += 1;
                this.col = 0;
                return this.newToken("\n", TokenType.Newline)
            }
        }else{
            return this.newToken("eof", TokenType.EOF)
        }

        if(this.skipWhitespace()) return this.newToken("eof", TokenType.EOF);

        const first_character = this.char();
        const token = this.newToken(first_character);

        switch(first_character){
            case "\"":
                token.value = this.get_characters_while((x)=> x != "\"").join("");
                this.advance()
                token.kind = TokenType.LiteralString;
                break;
            case "#":
                this.get_characters_while((x)=> x != "\n");
                this.advance()
                return this.next()
            case ">": 
                token.kind = (this.peek() == "=" ? TokenType.LessThanEqual : TokenType.LessThan);
                token.value += this.char();
                break;
            case "<":
                token.kind = (this.peek() == "=" ? TokenType.GreaterThanEqual : TokenType.GreaterThan);
                token.value += this.char();
                break;
            case "=":
                token.kind = (this.peek() == "=" ? TokenType.DoubleEqual : TokenType.SingeEqual);
                token.value += this.char();
                break;
            case "!":
                token.kind = (this.peek() == "=" ? TokenType.NotEqual : TokenType.ExclamationMark);
                token.value += this.char();
                break;
            case "(": token.kind = TokenType.Lparen; break;
            case ")": token.kind = TokenType.Rparen; break;
            case ";": token.kind = TokenType.Semicolon; break;
            case ":": token.kind = TokenType.Colon; break;
            case "^": token.kind = TokenType.Hat; break;
            case "+": token.kind = TokenType.Plus; break;
            case ".": token.kind = TokenType.Dot; break;
            case ",": token.kind = TokenType.Comma; break;
            case "-":{
                if(!this.peek()){token.kind = TokenType.Dash; break}
                if(/[0-9]/.test(this.peek() as string)){
                    token.value += this.get_characters_while((x)=>/^\d*\.?\d*$/.test(x)).join("");
                    token.kind = token.value.includes(".") ? TokenType.LiteralFloat : TokenType.LiteralInt;
                }else{
                    token.kind = TokenType.Dash;
                }
                break;
            }
            case "/": token.kind = TokenType.SlashForward; break;
            case "*": token.kind = TokenType.Asterisk; break;
            default:
                //check for integer or float
                if(/[0-9]/.test(first_character)){
                    token.value += this.get_characters_while((x)=>/^\d*\.?\d*$/.test(x)).join("")
                    token.kind = token.value.includes(".") ? TokenType.LiteralFloat : TokenType.LiteralInt;
                    return token;
                }

                //if it wasnt that it is a keyword or identifier
                token.value += this.get_characters_while((x)=>/^[\w-]+$/.test(x)).join("")

                if(token.value == "true" || token.value == "false"){
                    token.kind = TokenType.LiteralBool;
                    return token;
                }
                
                token.kind = keywords.get(token.value) ?? TokenType.Identifier;

        }
        
        return token;
    }

    //returns the next token but does not advance the lexer
    peek_next(){
        const pos = this.pos;
        const row = this.row;
        const col = this.col;
        const result = this.next();
        this.pos = pos;
        this.row = row;
        this.col = col;
        return result;
    }
    
}